Switch terminal backend from Crossterm to Termina

the-mikedavis opened the pull request:
[Termina](https://github.com/helix-editor/termina) is a new terminal manipulation crate I've written that has a similar interface as Crossterm. The interface exposes lower-level details than Crossterm's, exposing escape sequences and pushing all of that handling to the application.

Termina has a minimal Windows integration that works because of [ConPTY](https://devblogs.microsoft.com/commandline/windows-command-line-introducing-the-windows-pseudo-console-conpty/) - we tell terminals on Windows to "speak VT" like a *NIX PTY would. This drops a lot of Windows-specific code for the old Console API and allows us to support VT features consistently on all platforms. In particular OSC52 bracketed paste should work on Windows now. The downside is that we will require Windows 10 like WezTerm: version 10.0.17763 or later. This is a fairly old version though, published ~Fall 2018. The Kitty keyboard protocol should work on WezTerm (when enabled in config) now, ~though it seems a bit buggy in my testing - needs some debugging~.

Termina also has an interface for polling and reading for arbitrary events and escape sequences. (This exists in Crossterm too but is private.) This allows us to detect VT extensions like synchronized rendering and true color support while setting up the terminal.

Rendering should be smoother on this branch since synchronized output sequences are enabled when we detect support for them (fixes #731). Termina also has better "compression" for SGR sequences (things like foreground/background/underline colors, bold, italic, etc.) similar to https://github.com/crossterm-rs/crossterm/pull/879 but covering all SGR attributes in addition to foreground and background - this should give the terminal less work to do when it parses the escapes we emit during rendering.

This needs some testing. I did some limited manual testing on Kitty and Ghostty on macOS, Kitty, Foot and WezTerm on Linux and cmd.exe, PowerShell and WezTerm on Windows.

Also fixes #12232 (the same way as #13223 did)

mariocala commented:
Wow that is a noticeable improvement 🚀

I have some issues with space+y copying though.

I'm using Ghostty (1.1.3) on MacOS (arm) and helix is running in a Zellij (0.41.1) session over SSH on a NixOS (arm) server.

Copying with the termcode (OSC52) provider randomly stops working (it is easily reproducible, when copying multiple times) and I sometimes get random text output in the buffer at the current selection which disappears when I (re)select the lines.

Also checked the main branch again and it still works as expected.

Let me know if you need further information.

RoloEdits commented:
I see you linked https://github.com/crossterm-rs/crossterm/pull/879, would the performance gains seen there benefit Windows this time now that the terminal is "spoken" to the same way on all platforms?

the-mikedavis commented:
> Copying with the termcode (OSC52) provider randomly stops working

I was able to reproduce this with the same setup. I changed the OSC52 code a bit to do some explicit locking and flushing and I haven't been able to reproduce it so far. That part is a bit hacky - we basically `print!()` the command instead of executing it on the terminal buffer. We would need a bigger refactor to find a way to fix this properly.

I also noticed that true-color detection was broken for me in that workflow (i.e. SSH with Ghostty. Probably I would need to install the Ghostty terminfo on the NixOS side). I've updated the true-color detection to use a query following https://github.com/termstandard/colors#querying-the-terminal so it works out of the box in this case.

> would the performance gains seen there benefit Windows this time...?

Crossterm does actually "speak VT," meaning that it emits escape sequences - like those used to set foreground/background colors - when it detects that ConPTY is available. The issue with its Windows handling is that it doesn't "hear VT": it enables `ENABLE_VIRTUAL_TERMINAL_PROCESSING` on the output handle (`CONOUT$` or stdout) but not `ENABLE_VIRTUAL_TERMINAL_INPUT` on the input handle (`CONIN$` or stdin) ([Microsoft docs](https://learn.microsoft.com/en-us/windows/console/high-level-console-modes)). So Crossterm was already using that optimization in https://github.com/crossterm-rs/crossterm/pull/879 on Windows as well. It lacked the ability to read VT inputs from ConPTY though, so we wouldn't ever get a response to request/response type escape codes like DECRQSS/DECRPSS (what I was mentioning in https://github.com/helix-editor/helix/pull/13224#issuecomment-2781513200 and what makes the true color thing above possible) or detecting support for the Kitty keyboard protocol or synchronized outputs.

RoloEdits commented:
Got some time to play around with this and I see some odd characters on the status bar (where I expect different ones) and when I press tab I see this:
![image](https://github.com/user-attachments/assets/3be791a0-a2b9-4d70-b681-e1c821e50ea9)
![image](https://github.com/user-attachments/assets/928c149f-58e4-4607-a463-8585a98e4aa1)

chtenb commented:
On windows 11 it hangs during startup on wezterm, but launches correctly on windows terminal. The logs don't show anything.

the-mikedavis commented:
The unicode stuff should be fixed in the latest push - I was accidentally resetting the input/output "code pages" (encodings) after querying for features. That should cover the weird glyphs for non-ascii characters and input problems like `:q!`. I'll spend some time on Windows today looking at the rest.

cotneit commented:
Awesome, everything I mentioned above aside from `conhost` is fixed now!

the-mikedavis commented:
What does your WezTerm config look like?

cotneit commented:
Updated WezTerm to latest nightly and the issue disappeared

Also tried stable `wezterm 20240203-110809-5046fc22` and it behaves same as conhost - hangs after `hx` is executed

So to sum it up:
- `wezterm 20250320-072107-a8735851` (latest nightly) - works
- `wezterm 20240203-110809-5046fc22` (stable) - hangs
- `wezterm 20250131-201646-b28bbbc4` - bug described above

My config is pretty basic:
```lua
return {
    color_scheme = 'tokyonight',
    default_prog = { 'pwsh.exe', '-NoLogo' },
}
```

cotneit commented:
Tried `conhost` and stable WezTerm after last commit, they no longer hang but noticed the following issues:
- No colors
- Same issue with CTRL/SHIFT/ALT/WIN/CAPSLOCK inserting spaces as described [above](https://github.com/helix-editor/helix/pull/13307#issuecomment-2794633700)

RoloEdits commented:
This is all with wezterm:

The text issue was fixed, but now no capabilities seem to have been detected
```log
2025-04-10T12:39:55.065 helix_tui::backend::termina [DEBUG] Detected terminal capabilities in 3.7429ms: Capabilities { kitty_keyboard: None, synchronized_output: false, true_color: false, extended_underlines: false }
```
With `allow_win32_input_mode` set to `false`, ESC no longer works, like it did previously (with the text glitching).

For me, capslock works as expected. ctrl and alt also seem to be fine. Colors are fine.

RoloEdits commented:
latest commit looks more promising with the capability check, though I wonder why true color and underlines are false?

```log
2025-04-10T13:32:29.740 helix_tui::backend::termina [DEBUG] Detected terminal capabilities in 2.708ms: Capabilities { kitty_keyboard: Some, synchronized_output: true, true_color: false, extended_underlines: false }
2025-04-10T13:32:29.753 helix_tui::backend::termina [DEBUG] The terminal fully supports the requested keyboard enhancement flags
```

the-mikedavis commented:
For better compatibility I switched to the foreground/background sequences suggested in the microsoft docs: https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences#extended-colors. (As I understand it it's not spec-wise correct but is an idiom used in the wild more often than the correct syntax.) That should fix the colors on conhost and stable WezTerm (which is quite old, probably it bundles an old conhost). It's the same format as crossterm uses so it should be the same compatibility-wise.

That last push should fix the detection: out of the box WezTerm should be `true` for synchronized output and `false` or `None` for everything else (or `Some` if you have the kitty keyboard flag enabled). It does support true color and underlines but it doesn't currently respond to a request that would tell us if it did (that would be https://github.com/wezterm/wezterm/pull/6856). Probably the log message should be clear that `false` actually means "no response to the query" rather than a definitive lack of support.

RoloEdits commented:
I did notice some weird behavior with escaping and scrolling as well as just in general with escape, in that pressing it seems to cause some kind of delay. Or like the press doesnt actually "happen" until I give another input? Kind of hard to show this, but you can see me scrolling when you see the rapid change in key inputs at the bottom right. A menu also pops up when all I am doing is scrolling. Also when the prompt is open I am spamming escape but nothing happens until I press, in this instance, an arrow key:
![termina](https://github.com/user-attachments/assets/d53456cb-d149-4b11-9fbd-20dba62ab39a)

the-mikedavis commented:
Is that with the Kitty keyboard protocol enabled? I've seen that behavior with WezTerm as well when it is enabled - termina doesn't seem to get anything from the input handle until another button is pressed - but I haven't debugged why yet. I don't see the same in Alacritty notably and I believe it has at least some support for the Kitty keyboard protocol.

RoloEdits commented:
> Is that with the Kitty keyboard protocol enabled?

That is correct, the behavior shown is with these set:
```lua
config.enable_kitty_keyboard = true
config.allow_win32_input_mode = true
```

the-mikedavis commented:
It looks like this is a bug in the Kitty keyboard encoding function in WezTerm. I sent a patch https://github.com/wezterm/wezterm/pull/6872. I'm not sure why the old encoding causes the key to 'queue up' and wait for a non-escape input but using the proper representation seems to fix it.

RoloEdits commented:
Ok, added those wezterm patches and now running them. 

```log
2025-04-11T22:34:56.878 helix_tui::backend::termina [DEBUG] Detected terminal capabilities in 2.5524ms: Capabilities { kitty_keyboard: Some, synchronized_output: true, true_color: true, extended_underlines: true }
2025-04-11T22:34:56.889 helix_tui::backend::termina [DEBUG] The terminal fully supports the requested keyboard enhancement flags
```

All capabilities seem to be detected. The keys and mouse seem to work correctly. I will run this more and get back if anything is up.

Thanks for getting those upstream fixes in, the time and effort you spent is greatly appreciated.

RoloEdits commented:
Took me a minute to figure out what I was feeling was off, kind of a subtle thing when you haven't found it: the undercurl seems to be resetting or being ignored till its updated.
![helix_termina_undercurl](https://github.com/user-attachments/assets/dbe5f095-faf3-4659-ba2c-c28fffe7742c)

chtenb commented:
> On windows 11 it hangs during startup on wezterm, but launches correctly on windows terminal. The logs don't show anything.

This seems to be resolved on the current tip.

the-mikedavis commented:
That underline behavior may be a bug with WezTerm, I haven't been able to reproduce on Kitty, Alacritty or Foot yet. I'll take a look at how WezTerm is parsing those underline commands

archseer commented:
I remember we were dealing with a wezterm bug like that before, I think it depended on the order of the style data

archseer commented:
Previously seen here: https://github.com/helix-editor/helix/pull/4061#issuecomment-1275407947

the-mikedavis commented:
I believe that was a different issue with the crossterm backend at the time since I could reproduce it then on different terminals. I added a commit to work around the problem. It seems like WezTerm doesn't like having the underline color included in the big SGR update that updates all attributes that changed between two cells. We can work around it for now by emitting a separate SGR update whenever the underline color changes.

j4james commented:
> It seems like WezTerm doesn't like having the underline color included in the big SGR update that updates all attributes that changed between two cells.

This is likely because of a parameter limit. WezTerm has a limit of 16 (at least last I tested), and attributes with subparameters count as multiple parameters. For example an RGB color sequence like `\e[38;2;255;128;64m` counts as 5 parameters, even though only one attribute is being changed.

16 is actually a fairly common limit, and it's not even the lowest! I believe MLTerm has a limit of 10. That's just enough for two RGB colors, but if you're trying to change other attributes at the same time they're liable to get dropped or corrupted. So you need to be very careful how much you change in one sequence.

the-mikedavis commented:
Aha, that makes sense, thanks! I'll update this escape to "chunk" the updates so they stay under that conservative limit. If we find lower limits we can tune that further.

the-mikedavis commented:
Looks like WezTerm just bumped its CSI param limit from 32 to 256: https://github.com/wezterm/wezterm/pull/6194. It's still probably better to keep the number of parameters low for wider compatibility though

cotneit commented:
I accidentally ran
`hx $nu.default-config-dir | path join 'zoxide.nu'`
instead of
`hx ($nu.default-config-dir | path join 'zoxide.nu')`

And something broke 

This is with `nushell 0.103.0` in Windows Terminal

master:
![image](https://github.com/user-attachments/assets/2d8cb105-f999-4ddf-aa5b-7551636d022b)

This branch:
![image](https://github.com/user-attachments/assets/47df3026-55cc-4ab5-a7fc-02e4f74c5966)

the-mikedavis commented:
That should be fixed now after https://github.com/helix-editor/termina/commit/bd391b13fa19e22c9d09a782928491029a202bf4. Also added the "chunking" for SGR attributes

RoloEdits commented:
Checked to see if there is anything obvious in the samples for `termina`, and given that the edges are all pointers to the windows kernel(I believe), seems this is working great as a wrapper for stdio.

![image](https://github.com/user-attachments/assets/21df0bc6-8773-4a9d-8997-f0ebd667ec7f)

RoloEdits commented:
For those building from source, given that there is no `rev` to lock to a commit, you need to add the `..` pattern to `WindowSize` for the pixel info, as this is whats on `master`.

Helix itself does not use them, as stated here https://github.com/helix-editor/termina/pull/6#pullrequestreview-2844241799

mariocala commented:
> My cursor is disappearing when I move from right to left over a closing bracket via the `h` key in normal mode.
> 
> This only happens with Ghostty+Tmux (not standalone Ghostty). Can't reproduce on (helix) main branch.
> 
>  Screen.Recording.2025-05-19.at.22.14.22.mov

I think this has something to do with chunking in termina. Decreasing the [`parameter_chunk_size`](https://github.com/helix-editor/termina/blob/a302bf2b6d8814e6c3ff92a233c392808e5c2a59/src/escape/csi.rs#L386) to 9 makes things more broken while increasing it to 11 fixes all **apparent** issues for me. Still not sure what the actual problem is though..

the-mikedavis commented:
I will look at cutting the scope of this a bit so that we can get this merged down but so that we don't start enabling many other features. The querying system seems brittle over SSH in my testing: sometimes the responses are printed by the terminal before Helix has a chance to handle them. I guess there must be some race between enabling raw mode and the queries but I haven't looked into it very far yet.

RoloEdits commented:
One thing I wondered about was if this would address any of the unicode issues mentioned in https://github.com/helix-editor/helix/issues/6012?

the-mikedavis commented:
This PR shouldn't affect grapheme width: it's an unfortunate mismatch that a terminal application (Helix) can have a different understanding of width than a terminal emulator. There's recent work on this problem in https://sw.kovidgoyal.net/kitty/text-sizing-protocol/ which looks promising. And it should be easy to add that protocol to Termina. It's quite young though so I'd like to see other programs - terminal applications or emulators - adding support before we jump in. Currently it would only benefit Kitty users.

j4james commented:
> This PR shouldn't affect grapheme width: it's an unfortunate mismatch that a terminal application (Helix) can have a different understanding of width than a terminal emulator.

Just on this point, note that there's a mode you can query (`?2027`) that will tell you whether or not a terminal supports advanced Unicode features like grapheme clustering, which can affect the glyph width measurements. There is nice article on the subject here:
https://mitchellh.com/writing/grapheme-clusters-in-terminals

Last I checked there were around a half dozen terminals supporting it, which is not a lot, but generally you can assume that those not supporting it are going to be using the old wcwidth algorithm. So as a rule of thumb, if the mode is set you use the advanced Unicode width measurements, and if not set (or not supported) you'd assume a simple wcwidth algorithm.

It's not going to be perfect, but it's better than nothing.

the-mikedavis commented:
I haven't tested it myself but probably not - Termina's parser is a slightly modified version of Crossterm. I would need to do a deeper dive into what escape codes show up for AltGr keys to understand what changes Termina needs.

aschey commented:
> I haven't tested it myself but probably not - Termina's parser is a slightly modified version of Crossterm. I would need to do a deeper dive into what escape codes show up for AltGr keys to understand what changes Termina needs.

I [looked into this a while back](https://github.com/crossterm-rs/crossterm/pull/937#issuecomment-2618083519). There are some hacky workarounds used by Wezterm and Windows Terminal that will likely need to be replicated, unfortunately.

fnuttens commented:
@the-mikedavis @aschey FYI I just tried termina on my Windows machine (WezTerm and Windows Terminal) and I was surprised to see that I reproduce none of the issues I previously had with AltGr bindings 🙂 

I saw a new problem with bindings relying on my dead key, though, but I’ll probably make a new issue to discuss that.

lianghu commented:
running hx from ssh, there is '$qm' symbol each time.   it's noticed after the terminal switch

```
lianghu@svr:~/$ hx test.c
$qmlianghu@svr:~/$
```

sicher commented:
> @the-mikedavis @aschey FYI I just tried termina on my Windows machine (WezTerm and Windows Terminal) and I was surprised to see that I reproduce none of the issues I previously had with AltGr bindings 🙂

Same here. Running WSL (Debian) in Windows terminal. #7129 seems to be fixed.

forivall commented:
I discovered that this change causes a regression in the behaviour described in #4668
